### PR TITLE
UpdateOrCreateToken get secrets err handling optimization

### PR DESCRIPTION
**Pull request by helen-frank**
Signed-off-by: helen <haitao.zhang@daocloud.io>

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
when UpdateOrCreateToken get secrets, if the error is not a NotFound, the error is thrown in time
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/phases/bootstraptoken/node/token.go [hunk 1]
```diff
@@ -43,6 +44,8 @@ func UpdateOrCreateTokens(client clientset.Interface, failIfExists bool, tokens
 		secret, err := client.CoreV1().Secrets(metav1.NamespaceSystem).Get(context.TODO(), secretName, metav1.GetOptions{})
 		if secret != nil && err == nil && failIfExists {
 			return errors.Errorf("a token with id %q already exists", token.Token.ID)
+		} else if err != nil && !apierrors.IsNotFound(err) {
+			return err
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/bootstraptoken/node/token.go:48]: can this be written like this?

```
	if err != nil && !apierrors.IsNotFound(err) {
		return err
	}	
	if secret != nil && failIfExists {
		return errors.Errorf("a token with id %q already exists", token.Token.ID)
	}
```

**Comment by neolit123** [comment]
> when UpdateOrCreateToken get secrets, if the error is not a NotFound, the error is thrown in time

can you explain in more detail what errors you are seeing and under what conditions?

File: cmd/kubeadm/app/phases/bootstraptoken/node/token.go (hunk 1 above)
Comment (helen-frank) [reply cmd/kubeadm/app/phases/bootstraptoken/node/token.go:48]: got it, thanks

**Comment by helen-frank** [comment]
> > when UpdateOrCreateToken get secrets, if the error is not a NotFound, the error is thrown in time
> 
> can you explain in more detail what errors you are seeing and under what conditions?

This is not a bug, this kind of treatment is to deal with the error in time, not to continue no matter what the error is.

**Review by neolit123 — COMMENTED** [review]
/approve

i think this is safe to change, if we don't retry the  API calls (wrap UpdateOrCreateTokens in a retry loop?)
i could not find any cases in the kubeadm code that do the retries.

**Comment by helen-frank** [comment]
> /approve
> 
> i think this is safe to change, if we don't retry the API calls (wrap UpdateOrCreateTokens in a retry loop?) i could not find any cases in the kubeadm code that do the retries.

Is a good idea, maybe I can add a concurrent processing to it, what do you think
If you think this is feasible, I can do this job as soon as possible.

**Comment by neolit123** [comment]
> Is a good idea, maybe I can add a concurrent processing to it, what do you think
If you think this is feasible, I can do this job as soon as possible.

no i think we should keep it the way it is, right now.
we don't have much concurrent code in kubeadm.

**Comment by helen-frank** [comment]
> > Is a good idea, maybe I can add a concurrent processing to it, what do you think
> > If you think this is feasible, I can do this job as soon as possible.
> 
> no i think we should keep it the way it is, right now. we don't have much concurrent code in kubeadm.

Got it, I will do it as soon as possible to retry the development of logic.

**Comment by neolit123** [comment]
> Got it, I will do it as soon as possible to retry the development of logic.

i meant we should probably keep it the way it is for now without retry.
if we decide to add retry to this code, ideally it should follow some of the existing retry patterns.

perhaps for a separate PR.

**Pull request merged**